In [122]:
import numpy as np 
import pandas as pd

In [137]:
df = pd.concat(map(pd.read_csv, ['Brazil.csv', 'germany.csv','germany1.csv', 
                                 'Portugal1.csv', 'Spain1.csv', 'Switzerland.csv',
                                'Switzerland1.csv', 'France1.csv',
                                'Brazil1.csv', 'Netherlands.csv', 
                                'Belgium.csv', 'USA.csv', 'England.csv',
                                'Argentina.csv', 'Uruguay.csv', 
                                'Senegal.csv', 'Ecuador.csv', 'CostaRica.csv',
                                'Mexico.csv', 'France.csv', 'Denmark.csv',
                                'Australia.csv', 'Serbia.csv',
                                'Cameroon.csv', 'Ghana.csv', 'Iran.csv',
                                'Poland.csv', 'SaudI.csv', 'Tunisia.csv', 
                                 'Japan.csv', 'Croatia.csv', 'Morocco.csv',
                                'USA1.csv', 'Spain2.csv', 'Portugal2.csv', 
                                 'England2.csv', 'Belgium2.csv', 'Uruguay2.csv',
                                'Canada2.csv', 'Netherlands2.csv', 'Saudi2.csv', 
                                'Croatia2.csv', 'Cameroon2.csv', 'Denmark2.csv',
                                'Costa2.csv', 'Morocco2.csv', 'Poland2.csv',
                                'Tunisia2.csv', 'Japan2.csv', 'Argentina2.csv', 
                                 'Italy.csv', 'South_Korea.csv', 'Nigeria.csv',
                                'Colombia.csv', 'Paraguay.csv', 'Russia.csv',
                                'Algeria.csv', 'Chile.csv', 'Yugoslavia.csv',
                                'Scotland.csv', 'Romania.csv', 'Greece.csv',
                                'Honduras.csv', 'Austria.csv', 'Bulgaria.csv']))



In [138]:
# 158 with wc 2022 teams

In [139]:
df.shape

(229, 17)

In [140]:
df.head()

goals_z      xg_z  crosses_z  boxtouches_z  passes_z  progpasses_z  \
0  0.423077  0.146923  -0.136154     -0.030000  0.429231      0.037692   
1  0.216923  0.348462   0.031538      0.158462  0.835385      0.626923   
2  0.113846  0.392308   0.599231      0.513846  0.833077      0.405385   
3  0.479231  0.609231   0.227692      0.450769  0.770769      0.042308   
4  0.877692  0.773846   0.428462      0.659231  0.754615      0.335385   

   takeons_z  progruns_z  tackles_z  interceptions_z  clearances_z  blocks_z  \
0   0.244615   -0.220000   0.216154         0.270000     -0.076923 -0.097692   
1   0.270000    0.266923   1.143846         0.834615     -0.059231 -0.134615   
2   0.521538    1.139231   1.109231         0.859231      0.084615  0.103077   
3   0.337692    0.927692   0.506923         1.015385      0.020769  0.381538   
4   0.023077    0.638462   0.493846         0.637692     -0.117692 -0.033846   

   aerials_z   fouls_z  fouled_z    nsxg_z  results  
0  -0.020000 -0.224615  0.100769 -0.124615        2  
1   0.196923 -0.016923  0.033846  0.146923        1  
2   0.683846  0.213846  0.353846  0.786923        9  
3   0.038462  0.039231  0.022308  0.692308        5  
4   0.572308 -0.016154 -0.096923  0.890769        5

In [141]:
df.corr()

goals_z      xg_z  crosses_z  boxtouches_z  passes_z  \
goals_z          1.000000  0.614471   0.083100      0.347052  0.180022   
xg_z             0.614471  1.000000   0.403588      0.677305  0.325280   
crosses_z        0.083100  0.403588   1.000000      0.642591  0.253538   
boxtouches_z     0.347052  0.677305   0.642591      1.000000  0.371557   
passes_z         0.180022  0.325280   0.253538      0.371557  1.000000   
progpasses_z     0.035536  0.115869   0.275480      0.241539  0.752023   
takeons_z        0.255771  0.359477   0.139355      0.232867  0.378437   
progruns_z       0.183783  0.386917   0.248537      0.368459  0.579869   
tackles_z        0.217050  0.372899   0.247140      0.395419  0.744986   
interceptions_z  0.173696  0.297896   0.225584      0.291525  0.744625   
clearances_z    -0.176944 -0.381781  -0.250359     -0.421810 -0.114533   
blocks_z         0.027814  0.027511   0.025173      0.109341  0.389846   
aerials_z        0.091151  0.082565   0.239659      0.259735  0.010501   
fouls_z         -0.096311 -0.019138   0.047913      0.013265  0.059154   
fouled_z         0.056096  0.180699   0.191122      0.180316  0.303039   
nsxg_z           0.380320  0.668377   0.672855      0.860818  0.488397   
results         -0.617758 -0.309682   0.026103     -0.198913 -0.143699   

                 progpasses_z  takeons_z  progruns_z  tackles_z  \
goals_z              0.035536   0.255771    0.183783   0.217050   
xg_z                 0.115869   0.359477    0.386917   0.372899   
crosses_z            0.275480   0.139355    0.248537   0.247140   
boxtouches_z         0.241539   0.232867    0.368459   0.395419   
passes_z             0.752023   0.378437    0.579869   0.744986   
progpasses_z         1.000000   0.155813    0.314251   0.524437   
takeons_z            0.155813   1.000000    0.555055   0.369867   
progruns_z           0.314251   0.555055    1.000000   0.519110   
tackles_z            0.524437   0.369867    0.519110   1.000000   
interceptions_z      0.559503   0.373391    0.413221   0.698235   
clearances_z        -0.034335  -0.094147   -0.056800  -0.207746   
blocks_z             0.360422   0.160701    0.276575   0.250302   
aerials_z            0.044744  -0.081984   -0.052188   0.069672   
fouls_z              0.201719   0.112818    0.035128   0.014739   
fouled_z             0.193694   0.332161    0.248921   0.302096   
nsxg_z               0.367694   0.395687    0.466431   0.488773   
results             -0.048934  -0.175708   -0.204465  -0.183665   

                 interceptions_z  clearances_z  blocks_z  aerials_z   fouls_z  \
goals_z                 0.173696     -0.176944  0.027814   0.091151 -0.096311   
xg_z                    0.297896     -0.381781  0.027511   0.082565 -0.019138   
crosses_z               0.225584     -0.250359  0.025173   0.239659  0.047913   
boxtouches_z            0.291525     -0.421810  0.109341   0.259735  0.013265   
passes_z                0.744625     -0.114533  0.389846   0.010501  0.059154   
progpasses_z            0.559503     -0.034335  0.360422   0.044744  0.201719   
takeons_z               0.373391     -0.094147  0.160701  -0.081984  0.112818   
progruns_z              0.413221     -0.056800  0.276575  -0.052188  0.035128   
tackles_z               0.698235     -0.207746  0.250302   0.069672  0.014739   
interceptions_z         1.000000     -0.119980  0.306129   0.021699  0.126177   
clearances_z           -0.119980      1.000000  0.239911   0.264881  0.204799   
blocks_z                0.306129      0.239911  1.000000   0.044825  0.187837   
aerials_z               0.021699      0.264881  0.044825   1.000000  0.107502   
fouls_z                 0.126177      0.204799  0.187837   0.107502  1.000000   
fouled_z                0.318093      0.036176  0.091365   0.027299  0.264862   
nsxg_z                  0.420159     -0.419707  0.150350   0.162866  0.053664   
results                -0.114134      0.079177 -0.070757  -0.031247  0.117869   

                 fo

In [142]:
df.shape

(229, 17)

In [143]:
df.duplicated().value_counts()

False    229
dtype: int64

In [144]:
df.duplicated().value_counts()

False    229
dtype: int64

In [145]:
df.to_csv('WC-stats.csv', index=False)


In [133]:
from sklearn.linear_model import LinearRegression
from sklearn.inspection import permutation_importance
from sklearn.model_selection import train_test_split
from sklearn.linear_model import Ridge



In [105]:
df.head()

goals_z      xg_z  crosses_z  boxtouches_z  passes_z  progpasses_z  \
0  0.423077  0.146923  -0.136154     -0.030000  0.429231      0.037692   
1  0.216923  0.348462   0.031538      0.158462  0.835385      0.626923   
2  0.113846  0.392308   0.599231      0.513846  0.833077      0.405385   
3  0.479231  0.609231   0.227692      0.450769  0.770769      0.042308   
4  0.877692  0.773846   0.428462      0.659231  0.754615      0.335385   

   takeons_z  progruns_z  tackles_z  interceptions_z  clearances_z  blocks_z  \
0   0.244615   -0.220000   0.216154         0.270000     -0.076923 -0.097692   
1   0.270000    0.266923   1.143846         0.834615     -0.059231 -0.134615   
2   0.521538    1.139231   1.109231         0.859231      0.084615  0.103077   
3   0.337692    0.927692   0.506923         1.015385      0.020769  0.381538   
4   0.023077    0.638462   0.493846         0.637692     -0.117692 -0.033846   

   aerials_z   fouls_z  fouled_z    nsxg_z  results  
0  -0.020000 -0.224615  0.100769 -0.124615        2  
1   0.196923 -0.016923  0.033846  0.146923        1  
2   0.683846  0.213846  0.353846  0.786923        9  
3   0.038462  0.039231  0.022308  0.692308        5  
4   0.572308 -0.016154 -0.096923  0.890769        5

In [106]:
y = df['results']

In [107]:
X = df.drop(columns=['results'])

In [108]:
X_train, X_test, y_train, y_test = train_test_split(
     X, y, test_size=0.20, random_state=42)

In [109]:
model = Ridge(alpha=1e-2).fit(X_train, y_train)
model.score(X, y)

0.42833709432272027

In [110]:
r = permutation_importance(model, X, y, n_repeats=30, random_state=0)

# Permutation feature importance

In [111]:
scoring = ['r2', 'neg_mean_absolute_percentage_error', 'neg_mean_squared_error']
r_multi = permutation_importance(
     model, X, y, n_repeats=30, random_state=0, scoring=scoring)

for metric in r_multi:
     print(f"{metric}") 
     r = r_multi[metric]
     for i in r.importances_mean.argsort()[::-1]:
            if r.importances_mean[i] - 2 * r.importances_std[i] > 0:
                print(f"    {X.columns[i]:<8}"
                      f"{r.importances_mean[i]:.3f}"
                      f" +/- {r.importances_std[i]:.3f}")
                
                

r2
    goals_z 0.922 +/- 0.083
    progruns_z0.073 +/- 0.022
    xg_z    0.071 +/- 0.024
    fouled_z0.044 +/- 0.013
    boxtouches_z0.029 +/- 0.013
    crosses_z0.028 +/- 0.014
    passes_z0.026 +/- 0.011
neg_mean_absolute_percentage_error
    goals_z 0.672 +/- 0.080
    fouled_z0.060 +/- 0.022
neg_mean_squared_error
    goals_z 71.350 +/- 6.456
    progruns_z5.680 +/- 1.669
    xg_z    5.500 +/- 1.852
    fouled_z3.377 +/- 0.988
    boxtouches_z2.245 +/- 1.026
    crosses_z2.158 +/- 1.067
    passes_z1.983 +/- 0.886


- df.shape = (158, 17)
- Ridge Regression = 0.42629810615686115
- r2
    - goals_z 1.282 +/- 0.149
    - progruns_z0.096 +/- 0.029
    - xg_z    0.095 +/- 0.030
    - takeons_z0.026 +/- 0.013
- neg_mean_absolute_percentage_error
    - goals_z 0.988 +/- 0.115
- neg_mean_squared_error
    - goals_z 108.981 +/- 12.672
    - progruns_z8.139 +/- 2.494
    - xg_z    8.047 +/- 2.509
    - takeons_z2.195 +/- 1.067


- df.shape = (125, 17)

- Ridge Regression = 0.4494483686544971

- r2
    - goals_z 0.982 +/- 0.123
    - progruns_z0.132 +/- 0.039
    - xg_z    0.056 +/- 0.017
    - boxtouches_z0.046 +/- 0.023
    - nsxg_z  0.039 +/- 0.013
    - aerials_z0.031 +/- 0.014
- neg_mean_absolute_percentage_error
    - goals_z 0.888 +/- 0.122
    - progruns_z0.181 +/- 0.059
- neg_mean_squared_error
    - goals_z 89.125 +/- 11.206
    - progruns_z11.964 +/- 3.505
    - xg_z    5.055 +/- 1.543
    - boxtouches_z4.180 +/- 2.062
    - nsxg_z  3.509 +/- 1.182
    - aerials_z2.782 +/- 1.244

- df.shape = (145, 17)
- Ridge Regression = 0.43985733944194594

- r2
    - goals_z 0.919 +/- 0.122
    - progruns_z0.077 +/- 0.028
    - nsxg_z  0.069 +/- 0.026
    - boxtouches_z0.034 +/- 0.015
- neg_mean_absolute_percentage_error
    - goals_z 0.817 +/- 0.115
    - progruns_z0.124 +/- 0.036
    - tackles_z0.060 +/- 0.022
    - boxtouches_z0.048 +/- 0.023
- neg_mean_squared_error
    - goals_z 78.467 +/- 10.375
    - progruns_z6.549 +/- 2.364
    - nsxg_z  5.854 +/- 2.186
    - boxtouches_z2.943 +/- 1.292

- NMAPE = the higher the better
- NMSE
- R2 =  is a statistical measure of how close the data are to the fitted regression line . It is the percentage of variation (from 0 to 1) explained by the relationship between two variables.. metric that is used to evaluate the performance of a regression-based machine learning model.